xmtp_mls/src/worker.rs

```diff
@@ -0,0 +1,130 @@
+use crate::{configuration::WORKER_RESTART_DELAY, groups::device_sync::worker::SyncMetric, Client};
+use metrics::WorkerMetrics;
+use parking_lot::Mutex;
+use std::{fmt::Debug, hash::Hash, marker::PhantomData, sync::Arc};
+use xmtp_api::XmtpApi;
+use xmtp_db::XmtpDb;
+
+pub mod metrics;
+
+#[derive(PartialEq, Eq, Hash)]
+pub enum WorkerKind {
+    DeviceSync,
+    DisappearingMessages,
+    KeyPackageCleaner,
+}
+
+pub(crate) trait WorkerManager: Send + Sync {
+    fn sync_metrics(&self) -> Option<Arc<WorkerMetrics<SyncMetric>>>;
+    fn spawn(&self);
+}
+
+impl<Core, Metric> WorkerManager for WorkerRunner<Core, Metric>
+where
+    Core: Send + Sync,
+    Metric: Send + Sync + 'static,
+{
+    fn sync_metrics(&self) -> Option<Arc<WorkerMetrics<SyncMetric>>> {
+        if std::any::TypeId::of::<Metric>() == std::any::TypeId::of::<SyncMetric>() {
+            self.metrics.as_ref().lock().clone().map(|arc| {
+                // This is safe because we verified Metric == SyncMetric
+                unsafe { std::mem::transmute::<Arc<WorkerMetrics<Metric>>, Arc<WorkerMetrics<SyncMetric>>>(arc.clone()) }
+            })
+        } else {
+            None
+        }
+    }
+
+    fn spawn(&self) {
+        (self.spawn_fn)();
+    }
+}
+
+pub struct WorkerRunner<Core, Metric = NoMetric> {
+    pub metrics: Arc<Mutex<Option<Arc<WorkerMetrics<Metric>>>>>,
+    spawn_fn: Box<dyn Fn() + Send + Sync>,
+    _core: PhantomData<Core>,
+}
+
+impl<Core, Metric> WorkerRunner<Core, Metric>
+where
+    Metric: PartialEq + Hash + Send + Sync + 'static,
+{
+    pub fn register_new_worker<ApiClient, Db>(client: &Client<ApiClient, Db>)
+    where
+        ApiClient: XmtpApi + Send + Sync + 'static,
+        Db: XmtpDb + 'static,
+        Core: Worker<ApiClient, Db, Metric> + 'static,
+    {
+        let metrics = Arc::new(Mutex::default());
+        let runner = WorkerRunner {
+            metrics: metrics.clone(),
+            spawn_fn: Box::new({
+                let client = client.clone();
+                let metrics = metrics.clone();
+                move || {
+                    Self::spawn_worker_internal(&client, &metrics);
+                }
+            }),
+            _core: PhantomData::<Core>,
+        };
+
+        runner.spawn();
+        let kind = Core::kind();
+        let runner = Box::new(runner);
+        client.context.workers.lock().insert(kind, runner as Box<_>);
+    }
+
+    pub(crate) fn spawn_worker_internal<ApiClient, Db>(
+        client: &Client<ApiClient, Db>,
+        metrics: &Arc<Mutex<Option<Arc<WorkerMetrics<Metric>>>>>,
+    ) where
+        ApiClient: XmtpApi + 'static,
+        Db: XmtpDb + Send + Sync + 'static,
+        Core: Worker<ApiClient, Db, Metric> + 'static,
+    {
+        let mut core = Core::init(client);
+        *metrics.lock() = core.metrics();
```

Comment:
are we overwriting the metrics here/I'm having a hard time understanding how the metrics connect back to each worker here